### PR TITLE
Upgrade terraform-ls to 0.29.2

### PR DESCRIPTION
**Pull request by rchl**

### CODE REVIEW

File: LSP-terraform.sublime-settings [hunk 1]
```diff
@@ -5,20 +5,12 @@
...
   ],
   "selector": "source.terraform | source.terraform-vars",
-  // File watching functionality only works with "LSP-file-watcher-chokidar" package installed.
-  "file_watcher": {
-    "patterns": [
-      "**/*.tf",
-      "**/*.tfvars"
-    ],
-    "events": ["create", "change", "delete"]
-  },
```
Comment (rchl) [inline LSP-terraform.sublime-settings:15]: @radeksimko I can see that now watching is registered dynamically but the registered patterns don't include those. Is that on purpose?

Comment (radeksimko) [reply LSP-terraform.sublime-settings:15]: @rchl The assumption is that clients always send changes in `**/*.tf` and `**/*.tfvars`. If there was a way to communicate these statically in `initialize` response, the server would do so, but unfortunately there isn't, so we just document it. The dynamic registration itself is gated on a capability, and so we only use it for non-essential files.

TL;DR clients should continue sending `**/*.tf` and `**/*.tfvars`.

File: LSP-terraform.sublime-settings [hunk 2]
```diff
@@ -5,20 +5,12 @@
...
   "initializationOptions": {
-    "rootModulePaths": [],
-    "excludeModulePaths": [],
     "experimentalFeatures.validateOnSave": false,
     "experimentalFeatures.prefillRequiredFields": false,
-    "ignoreDirectoryNames": [],
     "ignoreSingleFileWarning": false,
+    "indexing.ignoreDirectoryNames": [],
+    "indexing.ignorePaths": [],
+    "terraform.logFilePath": "",
```
[the remark this answers is not in the document]
Comment (rchl) [reply LSP-terraform.sublime-settings:14]: The dots will be expanded into a dict so from the server perspective it will be as expected.

We supported "dotted" properties to match vscode preferences like https://github.com/hashicorp/vscode-terraform/blob/c1acf84bae1479c1eadcbc722b47fe3cdf028818/package.json#L369-L377

File: LSP-terraform.sublime-settings (hunk 1 above)
Comment (rchl) [reply LSP-terraform.sublime-settings:15]: the clients will send changes in those when those are manually changed in the editor but without file watcher we won't be notifying when something external changes those. If the latter is expected then the server should also dynamically set up watcher for those.

File: LSP-terraform.sublime-settings (hunk 2 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply LSP-terraform.sublime-settings:14]: Got it - that's fine then. Didn't know you do this magic expansion too. 😄 

There are two reasons we use the dotted-format in the VS Code extension:

1. The settings UI does wouldn't render objects so users would be forced to re-configure by hand-editing JSON.
2. The settings UI uses the dots as a way to group/separate the different options.

Feel free to do whatever feels appropriate in terms of UX for Sublime Text [users] here.

[the remark this answers is not in the document]
Comment (rchl) [reply LSP-terraform.sublime-settings:14]: This way is fine for ST also. It's better in terms of schema and descriptions of settings.

File: LSP-terraform.sublime-settings (hunk 1 above)
Comment (radeksimko) [reply LSP-terraform.sublime-settings:15]: Right, we could register those patterns dynamically too, I'm just not entirely sure how different clients handle duplicate patterns. If we decide to _only_ use dynamic registration for all patterns that brings a question of whether we leave some clients behind, just because they don't support dynamic registration, but they do support file watching. The LSP spec is not ideal in that respect.

Feel free to raise that idea as [an issue](https://github.com/hashicorp/terraform-ls/issues/new/choose).

Comment (rchl) [reply LSP-terraform.sublime-settings:15]: I'm not sure I see an issue with using only dynamic registration.
If the client doesn't support dynamic registration then it can keep using a manual watcher like we did here.
It doesn't feel right to register some patterns through dynamic registration and for some other rely on the client knowing what to watch.

Comment (radeksimko) [reply LSP-terraform.sublime-settings:15]: Dynamic only registration works great as long as the clients support it. Dynamic registration is _optional_ part of LSP, unlike file watching itself, so it is implied that clients can send `textDocument/didChangeWatchedFiles` notifications to the server, but it's impossible for the server to tell clients what these files should be without using **optional** part of the protocol (i.e. dynamic registration) - which feels wrong from the protocol design perspective.

I would prefer static registration for all patterns, but we cannot assume all client maintainers will read the docs. Also in our case most patterns are highly unlikely to change at runtime, so it seems unfortunate to have to do the extra roundtrip for this. The only reason we chose to use dynamic registration for the other patterns is for the complexity and amount of patterns, which is hard to convey via docs.

That said, given that dynamic is the only way in LSP, I'd be fine with moving it all there, as long as we can confirm that the majority of clients actually implement dynamic registration. I don't know if that's true as I haven't had the time to do that research yet. The one downside of doing so is missing some change notifications between `initialize` and `initialized`.

Until it is implemented though the assumption is that all clients _always_ send `**/*.tf` and `**/*.tfvars`.

Comment (rchl) [reply LSP-terraform.sublime-settings:15]: > Dynamic registration is _optional_ part of LSP, unlike file watching itself,

We've discussed this before so I might be repeating myself but my stance here is that before it was possible to dynamically register file watchers, it was client's responsibility to know whether server needs to receive notifications or not. So it wasn't very LSP-like and it was basically wrong.

Since dynamic registration it's now possible to handle it properly by telling clients what to watch.

If client is setting up the watcher the old way, it's the client's responsibility to either keep doing or remove it on updating to new server version. Unless the client can't control the server version, the it has a problem.

> The one downside of doing so is missing some change notifications between `initialize` and `initialized`.

I don't think that this should be an issue as I would expect the server to start reading/initializing the project just after sending the `initialized`. So it would wouldn't need to be notified about changes before that.

Comment (rchl) [reply LSP-terraform.sublime-settings]: restored manual watcher configuration for now
